rfc/pipeline-builder.md

```diff
@@ -0,0 +1,234 @@
+// ... any other calls needed to build the stage
+    .onFailure(FailureStrategy.HALT_BRANCH_AND_FAIL_PIPELINE)
+    .build();
+```
+
+### Dependencies
+
+The `pipeline-builder` library makes heavy use of two Java libraries to generate pipelines, [Jackson](https://github.com/FasterXML/jackson) and [Lombok](https://projectlombok.org/).
+Jackson provides an easy way to transform data classes to JSON, and Lombok uses annotations on these data classes to generate methods at compile time to make the creation of pipeline-related objects as simple as possible.
+
+**note:** In this context, _data classes_ are class definitions that describe objects holding values. They do not perform any sort of computation on this data and their objects are generally immutable once constructed. Most of the classes exported by the library follow this model.
+
+#### Lombok
+
+`pipeline-builder` relies on Lombok to generate the boilerplate code required to create data classes. This allows the library's code to focus on the data itself, and not on the code required to add getters, setters, builders, etc. In particular, the `@Builder` annotation is used with constructors that take all the object's properties as parameters, generating a builder class with one method per constructor parameter – each method having the same name as its corresponding parameter.
+
+## Drawbacks
+
+- The library is not a replacement for Deck, and cannot be used to edit pipelines in Spinnaker.
+- It requires some amount of knowledge of the internal format used by Spinnaker to represent pipelines and stages, especially when it comes to their inputs in the `context` map. This is somewhat mitigated by the availability of dedicated context classes for a few stage types, but more will need to be added in the future.
+- As Spinnaker evolves, the library will need to be updated to support new features, stage types, and their context classes.
+- No dependency exists on the Spinnaker codebase, so the library cannot automatically benefit from changes made to Spinnaker (more about this below).
+
+## Alternatives and Prior Art
+
+Some users already build and maintain their pipelines using other generators, often based on a different source language or on a templating language. Examples include [HCL](https://github.com/hashicorp/hcl) and [Jsonnet](https://jsonnet.org/) with [Sponnet](https://github.com/spinnaker/sponnet).
+
+While these tools help simplify the syntax of the pipeline definitions, they do not provide the same level of type safety, validation (with tests), refactoring support, and IDE compatibility as a library written in the same language as most of the Spinnaker codebase. They usually have some of the same drawbacks as the JSON format, in particular when it comes to poorly-named fields or the complex structures behind some common stage options.
+
+They are also not generally testable with code, unless these tests are limited to validating the generated JSON.
+
+## Evolution with Spinnaker
```

Comment:
Let me start by saying I'm a fan of this effort.  All by itself I expect it would be quite useful.  I'm curious if you/Apple have concrete plans to make the kinds of changes to orca that this section discusses.  Getting a "real" connection to spinnaker seems key to keeping it from getting stale.

I'm also curious if you considered an approach with jsonschema to define the structure of pipelines/various stages and then automated code generation from there.  I'm thinking about facilitating validation, both inside spinnaker itself, but also externally.  It's arguably a separate concern, but they're related.

Reply:
> I'm curious if you/Apple have concrete plans to make the kinds of changes to orca that this section discusses. Getting a "real" connection to spinnaker seems key to keeping it from getting stale.

The topic of interface-only modules is something we've talked about in the Spinnaker team, but more in the sense that we were lamenting the lack of a clear separation of interfaces and implementations (and wishing for this structure to exist) than as an actual plan to take this on. That said, I see two points to take into consideration here:

1: Extracting interfaces for the context classes could be done gradually, maybe adding them in batches for related stages for examples (one PR with all the Kubernetes types, the next one for stages related to artifacts, etc.)

2: The [existing context classes](https://github.com/nicolasff/pipeline-builder/tree/main/pipeline-builder/src/main/java/io/spinnaker/pipelinebuilder/json/contexts) that we already include do not always provide the exact same names for fields and methods as the names used in the actual `StageData` context class from Orca. This is because a major focus of this library is developer experience, and the representation used internally isn't always easy to understand. The existence of two "representations" for some fields – Java name vs JSON name – is something we would have to consider, just like we would have to ask whether it makes sense to have builder annotations on some classes and fields that we publish from an `orca-stages-api` module.

For example, take the (very basic) [context class](https://github.com/nicolasff/pipeline-builder/blob/9bf07ff40f3973d10c3be43b600e4e09054d3af8/pipeline-builder/src/main/java/io/spinnaker/pipelinebuilder/json/contexts/WaitContext.java#L40-L44) we created for the Wait stage:

```java
@Getter
@RequiredArgsConstructor(staticName="ofSeconds")
public class WaitContext implements ContextObject {
    @JsonProperty("waitTime") private final int waitTimeSeconds;
}
```

[This pipeline](https://github.com/nicolasff/pipeline-builder/blob/9bf07ff40f3973d10c3be43b600e4e09054d3af8/examples/src/main/java/com/apple/spinnaker/examples/ContextObjectsPipelineBuilder.java#L22-L26) shows how it is used with a Stage builder instead of passing a `Map` with `String` keys:

```java
        Stage waitStage = Stage.builder()
            .type(StageTypes.WAIT)
            .name("Wait a moment")
            .contextObject(WaitContext.ofSeconds(5))
            .build();
```
For reference, the approach without a dedicated class is to call `.context()` instead of `.contextObject`, [like this](https://github.com/nicolasff/pipeline-builder/blob/9bf07ff40f3973d10c3be43b600e4e09054d3af8/examples/src/main/java/com/apple/spinnaker/examples/TutorialPipelineBuilder.java#L23):

```java
            .context(Map.of("waitTime", 5))
```

This `WaitContext` class would be usable as a context class in Orca (maybe with some setters for Jackson), but as a consumer of `orca-stages-api` it would look a bit weird to have this static `ofSeconds` constructor in there. This is a simple example, but we could certainly have some that are significantly more complex.

---

> I'm also curious if you considered an approach with jsonschema to define the structure of pipelines/various stages and then automated code generation from there.

Not with jsonschema, no.

But some of the teams that use our Spinnaker instance manage their pipelines using [Pkl](https://pkl.apple.com/), a recently-released "configuration-as-code" language. They wrote a Pkl library with pre-defined types for common Spinnaker data structures, so users of this library build and maintain their pipelines in a type-safe, VCS-compatible, reviewable, and testable language.

These two solutions emerged independently to address the same problem; I thinks it says something about the usability of JSON, or at least our perception of JSON as a usable format for this purpose.

Reply:
All of this sounds great.  I'm a little nervous to use a new tool like Pkl, but the various attempts to use jsonschema over the years have failed and Spinnaker is sorely lacking validation like this.  So overall, I'd say yes please to all of it.

Extracting interfaces feels like it's big enough to be its own project.  Would you be willing to extract the first one, or perhaps write up some details on what to do / how to do it and then other folks could chip in to move that forward?